src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseFVMKernels.cpp

```diff
@@ -663,6 +663,356 @@ INST_FluxKernel( 5, FaceElementStencil );
+  // two-phase flow regime
+  else if( numMobilePhases == 2 )
+  {
+    // from Hui Cao's PhD thesis
+    localIndex const ip0 = mobilePhases[0];
+    localIndex const ip1 = mobilePhases[1];
+    real64 const dMob_dVolFrac[2] = { dPhaseRelPerm_dPhaseVolFrac[ip0][ip0] / phaseVisc[ip0],
+                                      -dPhaseRelPerm_dPhaseVolFrac[ip1][ip1] / phaseVisc[ip1] }; // using S0 = 1 - S1
+    real64 const denom = 1. / ( poreVol * ( mob[ip0] + mob[ip1] ) );
+    real64 const coef0 = denom * mob[ip1] / mob[ip0] * dMob_dVolFrac[ip0];
+    real64 const coef1 = -denom * mob[ip0] / mob[ip1] * dMob_dVolFrac[ip1];
+
+    phaseCFLNumber = fabs( coef0*phaseOutflux[ip0] + coef1*phaseOutflux[ip1] );
+  }
+  // three-phase flow regime
+  else if( numMobilePhases == 3 )
+  {
+    // from Keith Coats, IMPES stability: Selection of stable timesteps (2003)
+    real64 totalMob = 0.0;
+    for( localIndex ip = 0; ip < numMobilePhases; ++ip )
+    {
+      totalMob += mob[ip];
+    }
+
+    real64 f[2][2]{};
+    for( localIndex i = 0; i < 2; ++i )
+    {
+      for( localIndex j = 0; j < 2; ++j )
+      {
+        f[i][j]  = ( i == j )*totalMob - mob[i];
+        f[i][j] /= (totalMob * mob[j]);
+        real64 sum = 0;
+        for( localIndex k = 0; k < 3; ++k )
+        {
+          sum += dPhaseRelPerm_dPhaseVolFrac[k][j] / phaseVisc[k]
+                 * phaseOutflux[j];
+        }
+        f[i][j] *= sum;
+      }
+    }
+    phaseCFLNumber = f[0][0] + f[1][1];
+    phaseCFLNumber += sqrt( phaseCFLNumber*phaseCFLNumber - 4 * ( f[0][0]*f[1][1] - f[1][0]*f[0][1] ) );
+    phaseCFLNumber = 0.5 * fabs( phaseCFLNumber ) / poreVol;
+  }
+}
+
+
+template< localIndex NC >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void
+CFLKernel::
+  computeCompCFL( real64 const & poreVol,
+                  arraySlice1d< real64 const > compDens,
+                  arraySlice1d< real64 const > compOutflux,
+                  real64 & compCFLNumber )
+{
+
+
+  compCFLNumber = 0.0;
+  for( localIndex ic = 0; ic < NC; ++ic )
+  {
+    real64 const compMoles = compDens[ic] * poreVol;
+    if( compMoles > 1e-12 )
+    {
+      real64 const CFL = compOutflux[ic] / compMoles;
+      if( CFL > compCFLNumber )
+      {
+        compCFLNumber = CFL;
+      }
+    }
+  }
+}
+
+template< localIndex NC, localIndex NP >
+void
+CFLKernel::
+  launch( localIndex const size,
+          arrayView1d< real64 const > const & volume,
+          arrayView1d< real64 const > const & porosityRef,
+          arrayView2d< real64 const > const & pvMult,
+          arrayView2d< real64 const > const & compDens,
+          arrayView3d< real64 const > const & phaseRelPerm,
+          arrayView4d< real64 const > const & dPhaseRelPerm_dPhaseVolFrac,
+          arrayView3d< real64 const > const & phaseVisc,
+          arrayView2d< real64 const > const & phaseOutflux,
+          arrayView2d< real64 const > const & compOutflux,
+          real64 & maxPhaseCFLNumber,
+          real64 & maxCompCFLNumber )
```

Comment:
Would it also be of interest to compute and _store_ cell-wise CFL (so one could plot and analyze it in post-processing over the entire mesh)?

Reply:
Yes it makes sense to try to visualize the map of CFL numbers, I added this option (cell-wise CFL numbers are now stored).